Meta refactor

vcunat opened the pull request:
###### Motivation for this change

- Allow querying `meta` on disallowed derivations &ndash; broken, unfree, etc. But keep checking of the conditions including dependencies when instantiating a derviation.
- ~Better separate the code.~  Moved to #27405.

I was slightly taken aback by the fact that `nix-env -qa` does *not* check dependencies of packages, but I was unable to fix that without a significant performance hit, so I left it as it was.

I tested this fixes #21845, #22102, #7541, hopefully without any caveats.

Ericson2314 commented:
Yes please! As a bonus, pkgs/stdenv/generic/default.nix is actually readable now, thank god!

Ericson2314 commented on pkgs/stdenv/generic/default.nix:
```diff
@@ -184,15 +81,38 @@ let
-             in if !v.valid
-               then throwEvalHelp (removeAttrs v ["valid"])
-               else true;
+      pos' =
```
While we're at it, can we make this just the default arg for `pos`, rather than using `, pos ? null` and `if pos != null`?

vcunat replied:
Sounds good; pushed.

Ericson2314 commented on pkgs/stdenv/generic/default.nix:
```diff
@@ -244,37 +164,20 @@ let
+    in
+
+      lib.addPassthru
+        (derivation
```
Is the indentation right here? I think `derivation` is taking two arguments but they are indented at different levels.

vcunat replied:
No, the other argument is for `addPassthru`, and `derivation` takes a single argument only.

Ericson2314 replied:
Oh, hah. This seems so obvious in hindsight.

edolstra commented:
Does this mean that `nix-env -qa` will now show disallowed/broken packages? The intent was that they *don't* show up.

Moving meta checking into a separate file is a good idea, though I'm not sure if it should be added to `lib` (since `lib` is more for generic functions rather than specific functions used in only one place).

vcunat commented:
@edolstra: I left this as it was/is now, i.e. disallowed/broken packages are hidden but packages that *depend* on such are not.  Extending that would be easy and I had it so originally, but there would be a large performance price for that query.

I used lib, as I originally thought it would be used by `buildEnv` and perhaps others as well, but later I realized that they use stdenv, too... so currently I know no other place that would utilize the functions. If you like, we can move the file to `pkgs/stdenv/generic/check-meta.nix` or similar and not expose the functions.

Profpatsch commented:
I’m not sure if this preserves semantics.

Did you move the `assert`?

vcunat commented:
@Profpatsch: I don't get which `assert` exactly you mean, but the whole checking code *was* moved into a separate file.

Profpatsch commented:
@vcunat the one directly before `mkDerivation` in the old version.

vcunat commented:
@Profpatsch: if it's supposed to be this one
> \# Throw an error if trying to evaluate an non-valid derivation

that's now done inside `addMetaCheckInner`... and it's delayed to evaluating the name. (Your old PR did this when evaluating the builder, but that one's not evaluated at all during `nix-env -qa` by default.)

vcunat commented:
I suppose I'll have to look at the performance again.  For `nix-env -qa` it seemed OK but eval-release shows a significant performance hit:
```
$ time nix-instantiate --eval --strict --show-trace ./maintainers/scripts/eval-release.nix > /dev/null
36.16user 0.71system 0:24.42elapsed 150%CPU (0avgtext+0avgdata 2578940maxresident)k
```
```
$ time nix-instantiate --eval --strict --show-trace ./maintainers/scripts/eval-release.nix > /dev/null
46.40user 0.89system 0:32.41elapsed 145%CPU (0avgtext+0avgdata 3273876maxresident)k
```
(taken from the `tarball` job; similar to Hydra's evaluation)

Ericson2314 commented:
Anything this is waiting on? I'm eager to this this merged!

vcunat commented:
I haven't got to looking into the nontrivial performance regression.  There have been more pressing matters, even in nixpkgs.

Ericson2314 commented:
@vcunat Any time to look at this again? I think I'll need it for cross compilation soonish.

Ericson2314 commented on lib/check-meta.nix:
```diff
@@ -0,0 +1,121 @@
+      else if !allowBroken && meta.platforms or null != null && !lib.lists.elem system meta.platforms then
+        { valid = false; reason = "broken"; errormsg = "is not supported on ‘${system}’"; }
+      else { valid = true; };
+  # ^ checkMetaValidity
```
This comment probably isn't needed now this isn't in a huge scary file :).

0xABAB replied:
When does `null != null`?

Ericson2314 replied:
You have the precedence wrong, perhaps? Read it as `(meta.platforms or null) != null`. It's a shorter way of doing `meta ? platforms && meta.platforms != null`

vcunat replied:
Yes, the `or` operator binds even closer than function application.  https://github.com/NixOS/nix/issues/629

vcunat replied:
In any case, this discussion thread is on a stale diff.  Current PR is: https://github.com/NixOS/nixpkgs/pull/27045

vcunat commented:
@Ericson2314: I've been no good on nix* time in the past months an most of it was taken just by trying to fixup channels.  Still, I can't predict this well.

You're interested in the laziness part? (d2e4af7c7e ATM)  I don't yet know how exactly is cross-compilation evaluated now, even :-)

Ericson2314 commented:
@vcunat I suppose the laziness is useful too, but mainly I want it because stdenv.generic is big and scary and this makes it less so :). I had some vague thoughts too about separating stdenv from `mkDerivation`, and this tidyness would make thinking about that easier.

vcunat commented:
Right, that no-lazy part is easy to do and I found no performance impact, so I think I can do that before this week ends.

Ericson2314 commented:
Wonderful, thanks!

Ericson2314 commented:
@vcunat I really shouldn't be looking a gift horse in the mouth :), but when you do this, if you could use http://github.com/obsidiansystems/nixpkgs/tree/ericson2314-nixpkgs-base as your base branch, that would be *much* appreciated. That branch is my last merged PR to nixpkgs, so it always points to an upstream commit. That way, if master is broken, I know PR starts from a known-good change which includes all my cross changes on top. Then I can I can continue working from your PR tip rather than the merge commit and all will be well.

Alternatively, I'd be happy to pick up this PR if you are too busy---I think I should be able to figure out how to do the split without the laziness change. Certainly keeping the channels in a good state is important and tiresome work so you deserve a break!

vcunat commented:
OK, shouldn't be a problem, as I see it's not too late behind master.

0xABAB commented:
I think querying such information requires a database. One design is to let Nix become a database engine. Another (my idea) is to treat Nix as a data source and load that data into something which is already a database engine (don't really care which one) and then query that. Building the database index is an operation that should be done in the background. 

Can someone tell me why my idea would be a bad idea?

vcunat commented:
More like a DB that caches the evaluations done by nix.  Many would appreciate that and DBs seem common in packaging systems.  Still, I don't think it's really related to this PR.

vcunat commented:
First stage now on #27045.

vcunat committed:
stdenv meta checks: do them lazily

Allow querying meta on disallowed derivations – broken, unfree, etc.
But keep checking of the conditions including dependencies when
instantiating a derviation.

Performance: this causes creating another copy of the attr-map.
`nix-env -f . -qa` shows a roughly 2% increase in time and space.
`nix-instantiate --eval --strict --show-trace ./maintainers/scripts/eval-release.nix`
shows a higher increase (8-9%), but *real time* is not noticeably different,
probably due to GC being parallelized.

vcunat commented:
To be clear, `meta` and most other attributes can now (67e8392383) be queried on disallowed derivations.
